src/target/riscv/riscv-013.c

```diff
@@ -4394,182 +4394,309 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
 	return ERROR_OK;
 }
 
-static int write_memory_progbuf(struct target *target, target_addr_t address,
-		uint32_t size, uint32_t count, const uint8_t *buffer)
+static uint64_t get_from_buffer_and_log_write(target_addr_t address,
+		const uint8_t *buffer, uint32_t size)
 {
-	RISCV013_INFO(info);
+	assert(size <= 8);
+	uint64_t value = buf_get_u64(buffer, 0, 8 * size);
+	log_memory_access64(address, value, size, /*is_read*/ false);
+	return value;
+}
 
-	if (riscv_xlen(target) < size * 8) {
-		LOG_TARGET_ERROR(target, "XLEN (%d) is too short for %d-bit memory write.",
-				riscv_xlen(target), size * 8);
-		return ERROR_FAIL;
-	}
+static int write_abstract_arg_buf(struct target *target, target_addr_t address,
+		const uint8_t *buffer, uint32_t size)
+{
+	assert(size <= 8);
+	const uint64_t value = get_from_buffer_and_log_write(address, buffer, size);
+	return write_abstract_arg(target, /*index*/ 0, value, size > 4 ? 64 : 32);
+}
 
-	LOG_TARGET_DEBUG(target, "writing %d words of %d bytes to 0x%08lx", count, size, (long)address);
+/**
+ * This function is used to start the memory-writing pipeline.
+ * The pipeline looks like this:
+ * debugger -> dm_data[0:1] -> s1 -> memory
+ * Prior to calling it, the program buffer should contain the appropriate
+ * program.
+ * This function sets DM_ABSTRACTAUTO_AUTOEXECDATA to trigger second stage of the
+ * pipeline (dm_data[0:1] -> s1) whenever dm_data is written.
+ */
+static int write_memory_progbuf_startup(struct target *target, target_addr_t *address_p,
+		const uint8_t *buffer, uint32_t size)
+{
+	/* TODO: There is potential to gain some performance if the operations below are
+	 * executed as a single DMI batch.
```

Comment:
@en-sc Now that I read it again, is this what you actually meant?

```suggestion
	/* TODO: There is potential to gain some performance if the operations below are
	 * executed inside the first DMI batch (not separately).
```

Reply:
Yes. Thanks! Addressed.